### ISSUE TITLE
Reimplement spirv-tools assembler in spirv-tools-rs (or rspirv)

### ISSUE THREAD

**Issue by Jake-Shadle** [EmbarkStudios/rust-gpu#140]
As part of #117 I added an assembler type that wraps spirv-tools' assembler, however, I ran into a super strange issue that I still haven't found the root cause of.

Basically usage of some of C++'s (i)stringstream would work when being called from eg and example executable that statically linked the spirv-tools library (but not the c++ stdlib), but would, when part of the librustc_codegen_spirv shared library, result in, in the case of `std::stringstream` bad cast exception with a perfect ordinary uint32_t, or, in the case of `std::istringstream` fail to parse eg "42" and set the "bad" bit.

It is kind of frustrating as, AFAICT, the inputs would be the exact same (I even hashed them) going through the same calls, with the same options, and then one would work as intended and the other would fail, suggesting some kind of global state or possibly memory stomps? Anway, this only seemed to effect libstdc++, as windows and mac builds worked fine.

I finally just got fed up and replaced most of the number parsing in the assembler, see the [diff](https://github.com/KhronosGroup/SPIRV-Tools/compare/master...EmbarkStudios:patch-to-string) for what I ended up doing to work around this in most cases (need to support hexadecimals floats still).

Anyway, long story short, the spir-v assembly syntax is relatively straightforward, it shouldn't? be terribly difficult to replace the need for the spirv-tools' assembler, we can even take advantage of their excellent suite of tests.

**Comment by Jake-Shadle** [comment]
Then again, this is only used for testing, maybe it's not worth any effort.

**Comment by Jasper-Bekkers** [comment]
@Jake-Shadle rspirv already can assemble a textual representation; https://docs.rs/rspirv/0.7.0/rspirv/binary/trait.Disassemble.html

**Comment by Jake-Shadle** [comment]
Well, this is for assembling the binary from the text, which seems to also be implemented in rspirv via the Assemble trait, I was kind of confused why spirv-as was being used in the linker test, as presumably that would just be testing that rspirv is doing the correct thing? Or am I just missing something.

**Comment by Jake-Shadle** [comment]
I was confused, this isn't needed, `rspirv` should have everything we need.

**Comment by khyperia** [comment]
wait, @Jasper-Bekkers, that's Disassemble, not assembling (`spirv-as`)? @Jake-Shadle Is there another thing in rspirv that you found that supports creating a binary from text? I'm super confused, but I guess it doesn't matter if you figured it out :P

**Comment by Jake-Shadle** [comment]
Yes, I was confused as well by the use of `spirv-as` in the tests, but AFAICT the actual linking doing can [solely use rspirv](https://github.com/EmbarkStudios/rust-gpu/blob/9e67f83cd683e5776fac6b15b660713bd95e2e41/rustc_codegen_spirv/src/link.rs#L447-L452), no need for spirv-tools for it.

**Comment by khyperia** [comment]
Ah yeah - the use of spirv-as is so that we can write text format in tests, instead of opaque pre-assembled binaries.

**Comment by Jake-Shadle** [comment]
So yah, I guess we would want this then, at least for testing, but I'll keep closed for now since I doubt anyone will have time to work on it anytime soon.
